dtpc/doc/dtpcd.1
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DTPCD 1"
.TH DTPCD 1 "2022-11-03" "perl v5.30.0" "DTPC executables"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dtpcd \- DTPC daemon task for receiving and processing DTPC ADUs in bundles
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdtpcd\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBdtpcd\fR is a background \*(L"daemon\*(R" task that manages the reception and
processing of \s-1DTPC\s0 protocol data units.  It receives the payloads
of bundles destined for the \*(L"ipn\*(R"\-scheme endpoint whose node number is the
number of the local node and whose service number is the \s-1DTPC_RECV_SVC_NBR\s0
(129 as of the time of this writeng).
.PP
\&\s-1DTPC\s0 protocol data units are of two types: application data units (ADUs,
i.e., aggregations of application data items) and acknowledgments.  Each
acknowledgment is interpreted as authorization to release the buffer space
occupied by the node's local copy of the acknowledged \s-1ADU.\s0  Each \s-1ADU\s0 is
parsed into its constituent application data items, which are then delivered
to the applications awaiting them, and when required a \s-1DTPC\s0 end-to-end
acknowledgment \s-1PDU\s0 is returned to the \s-1DTPC PDU\s0 sender.
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
.ie n .IP "0" 4
.el .IP "\f(CW0\fR" 4
.IX Item "0"
\&\fBdtpcd\fR terminated normally, for reasons noted in the \fBion.log\fR file.  If
this termination was not commanded, investigate and solve the problem identified
in the log file and use \fBdtpcadmin\fR to restart \fBdtpcd\fR.
.ie n .IP "1" 4
.el .IP "\f(CW1\fR" 4
.IX Item "1"
\&\fBdtpcd\fR terminated abnormally, for reasons noted in the \fBion.log\fR file.
Investigate and solve the problem identified in the log file, then use
\&\fBdtpcadmin\fR to restart \fBdtpcd\fR.
.SH "FILES"
.IX Header "FILES"
No configuration files are needed.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
No environment variables apply.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
The following diagnostics may be issued to the \fBion.log\fR log file:
.IP "\s-1DTPC\s0 can't open own 'send' endpoint." 4
.IX Item "DTPC can't open own 'send' endpoint."
Bundle protocol agent has not been started.  See \fBion\fR\|(3).
.IP "dtpcd can't attach to \s-1DTPC.\s0" 4
.IX Item "dtpcd can't attach to DTPC."
\&\fBdtpcadmin\fR has not yet initialized \s-1DTPC\s0 protocol operations.
.SH "BUGS"
.IX Header "BUGS"
Report bugs to <ion\-dtn\-support@lists.sourceforge.net>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBdtpcadmin\fR\|(1), \fBion\fR\|(3)


release-notes.txt
= Release Notes for ION 4.1.1 =

November 19, 2021

%%%%%%%%%%%
= GENERAL =
%%%%%%%%%%%

The ION (interplanetary overlay network) software is a suite of communication
protocol implementations designed to support mission operation communications
across an end-to-end interplanetary network, which might include on-board
(flight) subnets, in-situ planetary or lunar networks, proximity links, deep
space links, and terrestrial internets.  Included in the ION software
distribution are the following packages:

- ici (interplanetary communication infrastructure), a set of libraries
  that provide flight-software-compatible support for functions on which
  the other packages rely, such as dynamic memory management, non-volatile
  storage management, and inter-task communication via shared memory.
  The ici libraries are designed to make the porting of IPN software to
  multiple operating systems - Linux, VxWorks, Solaris, etc. - as easy as
  possible. Ici now includes zco (zero-copy object), a library that 
  minimizes the copying of application data as it is encapsulated in 
  multiple layers of protocol structure while traversing the protocol
  stack.

- bp (bundle protocol), an implementation of the Delay-Tolerant
  Networking (DTN) architecture's Bundle Protocol.

- ltp (licklider transmission protocol), a DTN convergence layer protocol
  for reliable transmission over links characterized by long or highly
  variable delay.

- dgr (datagram retransmission), an alternative implementation of ltp
  that is designed for use over the Internet protocol stack.  dgr
  implements congestion control and is designed for relatively high
  performance.

- bssp (bundle streaming service protocol), a DTN convergence layer protocol
  for reliable transmission of streaming data.

- ams - an implementation of the CCSDS Asynchronous Message Service.

- cfdp - a class-1 (Unacknowledged) implementation of the CCSDS File
  Delivery Protocol.

- bss - a Bundle Streaming Service (BSS) for disruption-tolerant reliable
  data streaming.  BSS supports real-time streaming applications by
  passing the bundle payloads to the associated application for immediate
  display of the most recent data while storing all bundle payloads
  received into a database for user-directed playback.

- dtpc (delay-tolerant payload conditioning), an application framework
  providing TCP-like support for end-to-end retransmission and data delivery
  in transmission order without duplication or omission.

Features included:

- cgr - Contract graph routing: a method of dynamic routing designed for
  space based applications of ION, but still usable for terrestrial
  applications. It computes routes using scheduled communication and deals
  with time-varying network topology.

- brs - Bundle relay service: provides interconnectivity between networks
  that do not allow servers (those behind NAT for example). For more
  information, check man brsscla and man brsccla.

%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.1 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

Beginning with the ION 4.1.1 release, we will no longer be distributing 
two separate .tar.gz files (one that builds with BP version 6 by default and 
one that builds with BP version 7 by default). Instead, we will distribute a
single ION .tar.gz file that builds with BP version 7 by default. If you wish 
to use ION with BP version 6, you can use use the --enable-bpv6 configure 
option. If you are building ION from source on Windows for BP version 6, you
will need to run the command "make bpv6".

After ION version 4, all BP version 6 related changes will be bug fixes
unless explicitly identified as otherwise.

Branches included in this release:

misc-4.1.1-fixes
  * Improve the killm script 
  * Update pilot delay and add rate control for bpdriver
  * Change SourceCustodyRequired to NoCustodyRequested for AMS
  * Add a version of the req-003-multicast test for BP version 7
  * Fix a bug in librams.c that caused AMS petition assert/cancel
    flip-flop
  * Various updates for Makefiles (autoconf and manual)

apl-4.1.1-security-patches
  * Added BPSec Python test suite
  * Added accomodation of NULL terminator in security policy parameters
  * Added checks for NULL pointer references when a BCB targets an 
    extension block
  * Added TEST_POINT macros to BIB and BCB code to log test events

last-pre-git-tweaks
  * Add a warning about potential forwarding loops
  * Fix a typo in biberc documentation

autoconf-updates-MacOS_M1-WSL
  * Fix line ending encoding error in install_macos_sysctl.sh script
  * Add support for building ION on MacOS using the new M1 processor
    * At this time, ION works correctly on the M! processor with MacOS
      version 11.6 (Big Sur) but does not run properly on version
      12.0.1 (Monterey)
  * Add support to configure ION to build on WSL2 but not WSL1
  * ION has also now been tested on RedHat Linux


%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.0 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

ION 4.1.0 is the first release in which the BPv7 implementation is
conformant to the BPv7 protocol specification that will shortly be
published by IETF as a standards-track RFC.

The accompanying BPsec implementation is similarly conformant to
the BPSec specification that will also soon be a standard-track RFC.
NOTE, though, that the TCP convergence-layer protocol implementation
in ION 4.1.0 still conforms to the original TCPCL specification (RFC
7242), NOT to the upcoming TCPCL standards-track RFC; that upgrade
will be supported in a future release.

Another key change in this release is that Aggregate Custody Signaling
is no longer enabled by default.  (Due to some new features in ACS
configuration, when ACS is enabled the bpv6 bpclock daemon will no
longer run successfully unless the ACS database is initialized.
Since not all ION bpv6 users currently initialize ACS in node startup,
disabling ACS by default seems to be the best way forward.)  If you

dtpc/daemon/dtpcd.c
/*
	dtpcd.c:	DTPC daemon for transmission and reception.

	Authors: Giorgos Papastergiou, SPICE
		 Ioannis Alexiadis, SPICE

	Copyright (c) 2011, Space Internetworking Center,
	Democritus University of Thrace. ALL RIGHTS RESERVED.
										*/

#include <dtpcP.h>

typedef struct
{
	pthread_t	mainThread;
	int		running;
	BpSAP		rxSap;
	BpSAP		txSap;
} RxThreadParms;

/*	*	*	Receiver thread functions	*	*	*/

static void	*getBundles(void *parm)
{
	RxThreadParms		*parms = (RxThreadParms *) parm;
	char			ownEid[64];
	Sdr			sdr = getIonsdr();
	BpDelivery		dlv;
	uvast			profNum;
	Scalar			seqNum;
	char			type;
	unsigned int		aduLength;
	int			bytesRemaining;
	ZcoReader		reader;
	unsigned char		*buffer;
	int			bytesToRead;
	int			sdnvLength;
	unsigned char		*cursor;

	isprintf(ownEid, sizeof ownEid, "ipn:" UVAST_FIELDSPEC ".%d",
			getOwnNodeNbr(), DTPC_RECV_SVC_NBR);
	if (bp_open(ownEid, &(parms->rxSap)) < 0)
	{
		putErrmsg("DTPC can't open own 'recv' endpoint.", ownEid);
		parms->running = 0;
		return NULL;
	}

	writeMemo("[i] dtpcd receiver thread has started.");
	while (parms->running)
	{
		if (bp_receive(parms->rxSap, &dlv, BP_BLOCKING) < 0)
		{
			putErrmsg("dtpcd bundle reception failed.", NULL);
			parms->running = 0;
			continue;
		}

		switch (dlv.result)
		{
		case BpEndpointStopped:
			parms->running = 0;
			break;

		case BpPayloadPresent:
			CHKNULL(sdr_begin_xn(sdr));

			/* Since the max length of a Sdnv is 10 bytes,
			 * read 21 bytes to be sure that the Profile
			 * and Sequence number Sdnvs plus the type
			 * were read.					*/

			aduLength = zco_source_data_length(sdr, dlv.adu);
			bytesRemaining = aduLength;
			if (aduLength < 21)	/* Just in case we receive
						 * a very small adu.	*/
			{			
				bytesToRead = aduLength;
			}
			else
			{
				bytesToRead = 21;
			}

			buffer = MTAKE(bytesToRead);
			if (buffer == NULL)
			{
				putErrmsg("Out of memory.",NULL);
				return NULL;
			}

			cursor = buffer;
			zco_start_receiving(dlv.adu, &reader);
			if (zco_receive_headers(sdr, &reader, bytesToRead,
					(char *) buffer) < 0)
			{
				putErrmsg("dtpcd can't receive ADU header.",
						itoa(bytesToRead));
				sdr_cancel_xn(sdr);
				MRELEASE(buffer);
				parms->running = 0;
				continue;
			}

			type = *cursor;		/* Get the type byte.	*/
			cursor++;
			bytesRemaining--;
			sdnvLength = decodeSdnv(&profNum, cursor);
			cursor += sdnvLength;
			bytesRemaining -= sdnvLength;
			sdnvLength = sdnvToScalar(&seqNum, cursor);
			cursor += sdnvLength;
			bytesRemaining -= sdnvLength;

			/*	Mark remaining bytes as source data.	*/

			zco_delimit_source(sdr, dlv.adu, cursor - buffer,
					bytesRemaining);
			zco_strip(sdr, dlv.adu);
			MRELEASE(buffer);
			if (sdr_end_xn(sdr) < 0)
			{
				putErrmsg("dtpcd can't handle bundle delivery.",
						NULL);
				parms->running = 0;
				continue;
			}

			switch (type)
			{
			case 0x00:	/*	Received an adu.	*/
				switch (handleInAdu(sdr, parms->txSap, &dlv,
						profNum, seqNum))
				{
				case -1:
					putErrmsg("dtpcd can't handle inbound \
adu.", NULL);
					parms->running = 0;
					continue;

				case 1:
					if (parseInAdus(sdr) < 0)
					{
						putErrmsg("dtpcd can't parse \
							inbound adus.", NULL);
						parms->running = 0;
						continue;
					}

					/*	Intentional fall-through
					 *	to remaining cases.	*/

				case 0: 
				default:
					break;	/*	Inner switch.	*/
				}

				break;		/*	Middle switch.	*/

			case 0x01:	/*	Received an ACK.	*/
				if (handleAck(sdr, &dlv, profNum, seqNum) < 0)
				{
					putErrmsg("dtpcd can't handle ACK.",
							NULL);
					parms->running = 0;
					continue;
				}

				break;		/*	Middle switch.	*/
			default:
				writeMemo("[?] Invalid item type. Corrupted \
item?");
				break;		/*	Middle switch.	*/
			}

		default:
			break;			/*	Outer switch.	*/
		}

		bp_release_delivery(&dlv, 0);	/* We manually delete the
						 * ZCO elsewhere.	*/

		/*	Make sure other tasks have a chance to run.	*/

		sm_TaskYield();
	}

	bp_close(parms->rxSap);
	writeMemo("[i] dtpcd receiver has stopped.");
	return NULL;
}


/*	*	*	Main thread functions	*	*	*	*/

#if defined (ION_LWT)
int	dtpcd(saddr a1, saddr a2, saddr a3, saddr a4, saddr a5,
		saddr a6, saddr a7, saddr a8, saddr a9, saddr a10)
{
#else
int	main(int argc, char **argv)
{
#endif
	char		ownEid[64];
	BpSAP		txSap;
	RxThreadParms	parms;
	pthread_t	rxThread;

	if (bp_attach() < 0)
	{
		putErrmsg("DTPC can't attach to BP.", NULL);
		return 0;
	}

	isprintf(ownEid, sizeof ownEid, "ipn:" UVAST_FIELDSPEC ".%d",
			getOwnNodeNbr(), DTPC_SEND_SVC_NBR);
	if (bp_open_source(ownEid, &txSap, 1) < 0)
	{
		putErrmsg("DTPC can't open own 'send' endpoint.", ownEid);
		return 0;
	}

	if (txSap == NULL)
	{
		putErrmsg("dtpcd can't get Bundle Protocol SAP.", NULL);
		return 0;
	}

	if (dtpcAttach() < 0)
	{
		bp_close(txSap);
		putErrmsg("dtpcd can't attach to DTPC.", NULL);
		return 0;
	}

	parms.mainThread = pthread_self();
	parms.running = 1;
	parms.txSap = txSap;
	if (pthread_begin(&rxThread, NULL, getBundles,
		&parms,"dtpcd_receiver"))
	{
		bp_close(txSap);
		putSysErrmsg("dtpcd can't create receiver thread", NULL);
		return -1;
	}
	
	writeMemo("[i] dtpcd is running.");
	while (parms.running)
	{
		/*	Get an outbound ADU for transmission.		*/

		if (sendAdu(txSap) < 0)
		{
			writeMemo("[?] dtpcd can't dequeue outbound ADU; \
terminating.");
			parms.running = 0;
			continue;
		}
		
		/*	Make sure other tasks have a chance to run.	*/

		sm_TaskYield();
	}
	
	bp_interrupt(parms.rxSap);
	pthread_join(rxThread, NULL);
	bp_close(txSap);
	writeErrmsgMemos();
	writeMemo("[i] dtpcd has ended.");
	ionDetach();
	return 0;
}
